pkg/ebpf/telemetry/errors_collector_linux.go

```diff
@@ -53,43 +59,51 @@ func (e *EBPFErrorsCollector) Describe(ch chan<- *prometheus.Desc) {
 
 // Collect returns the current state of all metrics of the collector
 func (e *EBPFErrorsCollector) Collect(ch chan<- prometheus.Metric) {
-	e.T.mtx.Lock()
-	defer e.T.mtx.Unlock()
-
-	if e.T.helperErrMap != nil {
-		var hval HelperErrTelemetry
-		for probeName, k := range e.T.probeKeys {
-			err := e.T.helperErrMap.Lookup(&k, &hval)
-			if err != nil {
-				log.Debugf("failed to get telemetry for probe:key %s:%d\n", probeName, k)
-				continue
-			}
-			for index, helperName := range helperNames {
-				base := maxErrno * index
-				if count := getErrCount(hval.Count[base : base+maxErrno]); len(count) > 0 {
-					for errStr, errCount := range count {
-						ch <- prometheus.MustNewConstMetric(e.ebpfHelperErrors, prometheus.GaugeValue, float64(errCount), helperName, probeName, errStr)
-					}
+	e.T.Lock()
+	defer e.T.Unlock()
+
+	if !e.T.isInitialized() {
+		return // no telemetry to collect
+	}
+
+	e.T.forEachMapEntry(func(index telemetryIndex, val MapErrTelemetry) bool {
+		if count := getErrCount(val.Count[:]); len(count) > 0 {
+			for errStr, errCount := range count {
+				key := metricKey{
+					hash: index.key,
+					id:   mapErr,
+					err:  errStr,
 				}
+				delta := float64(errCount - e.lastValues[key])
+				if delta > 0 {
+					ch <- prometheus.MustNewConstMetric(e.ebpfMapOpsErrors, prometheus.CounterValue, delta, index.name, errStr)
```

Comment:
rather than creating a new metric each time, we could use `prometheus.NewCounterVec` once, and then `.WithLabelValues(index.name, errStr).Add(delta)`. See `perf_metrics.go` for an example.

Reply:
i was planning to change that in a separate PR, but I guess this change can be done together with the metric type change

Reply:
changed